File: frontend/src/types/advisory.ts
export type AdvisorySeverity = "low" | "medium" | "high";
export type AdvisorySubjectKind =
  | "crop"
  | "livestock"
  | "parcel"
  | "equipment"
  | "infrastructure"
  | "market"
  | "administrative"
  | "other";
export type AdvisorySubjectStatus = "planned" | "active" | "inactive" | "resolved";
export type AdvisoryReportCategory =
  | "pest"
  | "disease"
  | "weather_damage"
  | "drought"
  | "flood"
  | "input_shortage"
  | "market_bottleneck"
  | "infrastructure_damage"
  | "livestock_health"
  | "subsidy_or_payment"
  | "invasive_species"
  | "other";
export type AdvisoryReportStatus = "open" | "monitoring" | "resolved" | "escalated";
export type AdvisoryQueryCategory =
  | "production"
  | "plant_health"
  | "livestock_health"
  | "input_sourcing"
  | "regulatory"
  | "subsidy"
  | "market"
  | "weather"
  | "other";
export type AdvisoryQueryStatus = "open" | "answered" | "needs_follow_up";
export type AdvisoryMaterialStatus = "to_check" | "available" | "not_available" | "ordered";

export type AdvisorySubject = {
  id: string;
  name: string;
  kind: AdvisorySubjectKind;
  type?: string | null;
  location?: string | null;
  description?: string | null;
  quantity?: string | null;
  status?: AdvisorySubjectStatus | null;
  notes?: string | null;
};

export type AdvisoryMeasurement = {
  id: string;
  label: string;
  value: string;
  unit?: string | null;
  measured_at?: string | null;
  method?: string | null;
  location?: string | null;
  subject_ids: string[];
  report_ids: string[];
  query_ids: string[];
  notes?: string | null;
};

export type AdvisoryReport = {
  id: string;
  category: AdvisoryReportCategory;
  title: string;
  description?: string | null;
  status?: AdvisoryReportStatus | null;
  severity?: AdvisorySeverity | null;
  reported_at?: string | null;
  observed_at?: string | null;
  location?: string | null;
  recommended_follow_up?: string | null;
  subject_ids: string[];
  evidence_image_ids: string[];
  measurement_ids: string[];
};

export type AdvisoryQuery = {
  id: string;
  category: AdvisoryQueryCategory;
  question: string;
  status: AdvisoryQueryStatus;
  asked_at?: string | null;
  answer_summary?: string | null;
  source_urls: string[];
  subject_ids: string[];
  report_ids: string[];
  measurement_ids: string[];
  notes?: string | null;
};

export type AdvisoryMaterial = {
  id: string;
  name: string;
  purpose?: string | null;
  category?: string | null;
  status: AdvisoryMaterialStatus;
  supplier_name?: string | null;
  supplier_url?: string | null;
  subject_ids: string[];
  report_ids: string[];
  query_ids: string[];
  notes?: string | null;
};

export type AdvisoryRecordPayload = {
  version: "v2";
  title: string;
  profile_description?: string | null;
  default_location?: string | null;
  subjects: AdvisorySubject[];
  reports: AdvisoryReport[];
  queries: AdvisoryQuery[];
  measurements: AdvisoryMeasurement[];
  materials: AdvisoryMaterial[];
};

export type AdvisoryImageSourceKind = "upload" | "chat_attachment";

export type AdvisoryImageSummary = {
  id: string;
  case_id: string;
  chat_id?: string | null;
  attachment_id?: string | null;
  source_kind: AdvisoryImageSourceKind;
  name: string;
  mime_type?: string | null;
  byte_size: number;
  width: number;
  height: number;
  preview_url?: string | null;
  created_at: string;
  updated_at: string;
};

export type AdvisoryCaseSummary = {
  id: string;
  title: string;
  chat_id?: string | null;
  image_count: number;
  created_at: string;
  updated_at: string;
};

export type AdvisoryCaseDetail = AdvisoryCaseSummary & {
  default_location?: string | null;
  profile_description?: string | null;
  images: AdvisoryImageSummary[];
};

export type AdvisoryRecordResponse = {
  case_id: string;
  record: AdvisoryRecordPayload;
};

export type AdvisoryCaseDeleteResponse = {
  case_id: string;
  deleted: boolean;
};

export type AdvisoryImageUploadResponse = {
  image: AdvisoryImageSummary;
};

export type AdvisoryImageDeleteResponse = {
  case_id: string;
  image_id: string;
  deleted: boolean;
};

File: backend/app/schemas/advisory.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class AdvisorySchemaBase(BaseModel):
    model_config = ConfigDict(extra="forbid")


AdvisorySeverity = Literal["low", "medium", "high"]
AdvisorySubjectKind = Literal[
    "crop",
    "livestock",
    "parcel",
    "equipment",
    "infrastructure",
    "market",
    "administrative",
    "other",
]
AdvisorySubjectStatus = Literal["planned", "active", "inactive", "resolved"]
AdvisoryReportCategory = Literal[
    "pest",
    "disease",
    "weather_damage",
    "drought",
    "flood",
    "input_shortage",
    "market_bottleneck",
    "infrastructure_damage",
    "livestock_health",
    "subsidy_or_payment",
    "invasive_species",
    "other",
]
AdvisoryReportStatus = Literal["open", "monitoring", "resolved", "escalated"]
AdvisoryQueryCategory = Literal[
    "production",
    "plant_health",
    "livestock_health",
    "input_sourcing",
    "regulatory",
    "subsidy",
    "market",
    "weather",
    "other",
]
AdvisoryQueryStatus = Literal["open", "answered", "needs_follow_up"]
AdvisoryMaterialStatus = Literal[
    "to_check",
    "available",
    "not_available",
    "ordered",
]


class AdvisorySubject(AdvisorySchemaBase):
    id: str
    name: str
    kind: AdvisorySubjectKind
    type: str | None = None
    location: str | None = None
    description: str | None = None
    quantity: str | None = None
    status: AdvisorySubjectStatus | None = None
    notes: str | None = None


class AdvisoryMeasurement(AdvisorySchemaBase):
    id: str
    label: str
    value: str
    unit: str | None = None
    measured_at: str | None = None
    method: str | None = None
    location: str | None = None
    subject_ids: list[str] = Field(default_factory=list)
    report_ids: list[str] = Field(default_factory=list)
    query_ids: list[str] = Field(default_factory=list)
    notes: str | None = None


class AdvisoryReport(AdvisorySchemaBase):
    id: str
    category: AdvisoryReportCategory
    title: str
    description: str | None = None
    status: AdvisoryReportStatus | None = None
    severity: AdvisorySeverity | None = None
    reported_at: str | None = None
    observed_at: str | None = None
    location: str | None = None
    recommended_follow_up: str | None = None
    subject_ids: list[str] = Field(default_factory=list)
    evidence_image_ids: list[str] = Field(default_factory=list)
    measurement_ids: list[str] = Field(default_factory=list)


class AdvisoryQuery(AdvisorySchemaBase):
    id: str
    category: AdvisoryQueryCategory
    question: str
    status: AdvisoryQueryStatus = "open"
    asked_at: str | None = None
    answer_summary: str | None = None
    source_urls: list[str] = Field(default_factory=list)
    subject_ids: list[str] = Field(default_factory=list)
    report_ids: list[str] = Field(default_factory=list)
    measurement_ids: list[str] = Field(default_factory=list)
    notes: str | None = None


class AdvisoryMaterial(AdvisorySchemaBase):
    id: str
    name: str
    purpose: str | None = None
    category: str | None = None
    status: AdvisoryMaterialStatus = "to_check"
    supplier_name: str | None = None
    supplier_url: str | None = None
    subject_ids: list[str] = Field(default_factory=list)
    report_ids: list[str] = Field(default_factory=list)
    query_ids: list[str] = Field(default_factory=list)
    notes: str | None = None


class AdvisoryRecordPayload(AdvisorySchemaBase):
    version: Literal["v2"]
    title: str
    profile_description: str | None = None
    default_location: str | None = None
    subjects: list[AdvisorySubject] = Field(default_factory=list)
    reports: list[AdvisoryReport] = Field(default_factory=list)
    queries: list[AdvisoryQuery] = Field(default_factory=list)
    measurements: list[AdvisoryMeasurement] = Field(default_factory=list)
    materials: list[AdvisoryMaterial] = Field(default_factory=list)


class AdvisoryCaseSummary(AdvisorySchemaBase):
    id: str
    title: str
    chat_id: str | None = None
    image_count: int = Field(ge=0)
    created_at: str
    updated_at: str


class AdvisoryImageSummary(AdvisorySchemaBase):
    id: str
    case_id: str
    chat_id: str | None = None
    attachment_id: str | None = None
    source_kind: Literal["upload", "chat_attachment"]
    name: str
    mime_type: str | None = None
    byte_size: int = Field(ge=0)
    width: int = Field(ge=0)
    height: int = Field(ge=0)
    detailed_description: str | None = None
    location_label: str | None = None
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    preview_url: str | None = None
    created_at: str
    updated_at: str


class AdvisoryCaseDetail(AdvisoryCaseSummary):
    default_location: str | None = None
    profile_description: str | None = None
    images: list[AdvisoryImageSummary] = Field(default_factory=list)


class AdvisoryCaseCreateRequest(AdvisorySchemaBase):
    title: str


class AdvisoryCaseUpdateRequest(AdvisorySchemaBase):
    title: str | None = None


class AdvisoryRecordResponse(AdvisorySchemaBase):
    case_id: str
    record: AdvisoryRecordPayload


class AdvisoryRecordUpdateRequest(AdvisorySchemaBase):
    record: AdvisoryRecordPayload


class AdvisoryCaseDeleteResponse(AdvisorySchemaBase):
    case_id: str
    deleted: bool


class AdvisoryImageListResponse(AdvisorySchemaBase):
    case_id: str
    images: list[AdvisoryImageSummary] = Field(default_factory=list)


class AdvisoryImageUploadResponse(AdvisorySchemaBase):
    image: AdvisoryImageSummary


class AdvisoryImageDeleteResponse(AdvisorySchemaBase):
    case_id: str
    image_id: str
    deleted: bool
